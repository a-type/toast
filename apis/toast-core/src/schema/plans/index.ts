import { gql } from 'apollo-server-express';
import { Context } from 'context';
import { path, mergeDeepRight } from 'ramda';
import { id } from 'tools';

import * as meals from './meals';
import * as schedules from './schedules';

export const typeDefs = () => [
  gql`
    type Plan {
      id: ID!
    }

    extend type Query {
      plan: Plan
    }
  `,
  meals.typeDefs,
  schedules.typeDefs,
];

export const resolvers = [
  {
    Query: {
      plan: async (_parent, args, ctx: Context) => {
        const group = await ctx.graph.groups.getMine();
        let planId = path<string>(['planId'], group);

        if (!planId) {
          planId = id('plan');
          await ctx.graph.groups.mergeMine({ planId });
        }

        const plan = await ctx.firestore.plans.get(planId);

        return plan;
      },
    },
  },
  meals.resolvers,
  schedules.resolvers,
].reduce(mergeDeepRight, {});
