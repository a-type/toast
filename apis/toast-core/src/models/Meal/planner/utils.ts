import { ScheduleMeal } from 'models/Schedule/Schedule';
import { Meal, Schedule } from 'models';

export const MEALS_PER_DAY = 3;
export const DAYS_PER_WEEK = 7;

export const mealsFromGroceryDay = (schedule: Schedule): ScheduleMeal[] => {
  return [
    ...schedule.meals.slice(schedule.groceryDay * MEALS_PER_DAY),
    ...schedule.meals.slice(0, schedule.groceryDay * MEALS_PER_DAY),
  ];
};

const findCookedMeal = meal =>
  meal.mealIndex !== 0 &&
  ['LONG', 'MEDIUM', 'SHORT'].includes(meal.availability);

export const initializeWeek = (
  schedule: Schedule,
  startDateIndex: number,
  endDateIndex: number,
  findFirstMeal: (meal: ScheduleMeal) => boolean = findCookedMeal,
): Meal[] => {
  // skip breakfast for main planning
  // find first cook day after grocery day to use to initialize the week
  const firstCookMeal = mealsFromGroceryDay(schedule).find(findFirstMeal);

  if (!firstCookMeal) {
    throw new Error(
      "Sorry, we couldn't figure out where to start this plan based on your schedule.",
    );
  }

  const meals = new Array(1 + endDateIndex - startDateIndex)
    .fill(null)
    .map((_, dayIndex) =>
      new Array(MEALS_PER_DAY).fill(null).map((__, mealIndex) => {
        return Meal.createEmpty(startDateIndex + dayIndex, mealIndex);
      }),
    )
    .reduce((allMeals, dayMeals) => allMeals.concat(dayMeals), []);

  return meals;
};
