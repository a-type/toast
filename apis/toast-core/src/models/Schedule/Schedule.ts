import { id } from 'tools';
import Meal, { MealData } from 'models/Meal/Meal';
import planner from './planner';

export enum ScheduleStrategy {
  Basic = 'BASIC',
  Prep = 'PREP',
}

export type ScheduleMeal = {
  id: string;
  mealIndex: number;
  dayIndex: number;
  availability: string; // TODO: enum
  servings?: number;
};

type BaseScheduleData = {
  id?: string;
  defaultServings: number;
  strategy?: ScheduleStrategy;
  meals: ScheduleMeal[];
  warnings: string[];
};

export type ScheduleData = BaseScheduleData & {
  templateWeek?: MealData[];
};

type InternalScheduleData = BaseScheduleData & {
  templateWeek?: Meal[];
};

const createEmpty = (scheduleId): ScheduleData => {
  return {
    id: scheduleId,
    defaultServings: 2,
    strategy: ScheduleStrategy.Basic,

    meals: new Array(21).fill(null).map((__, mealIndex) => ({
      id: `${scheduleId}_m_${mealIndex}`,
      mealIndex,
      dayIndex: Math.floor(mealIndex / 3),
      availability: 'SKIP',
      servings: 2,
    })),

    templateWeek: [],

    warnings: [],
  };
};

export default class Schedule {
  data: InternalScheduleData;
  generateId: (name: string) => string;

  constructor(data?: ScheduleData, { generateId = id } = {}) {
    this.generateId = generateId;
    const incomingData = data || createEmpty(generateId('schedule'));
    this.data = {
      ...incomingData,
      templateWeek: incomingData.templateWeek.map<Meal>(meal => new Meal(meal)),
    };
  }

  get id() {
    return this.data.id;
  }

  get strategy() {
    return this.data.strategy;
  }

  get defaultServings() {
    return this.data.defaultServings;
  }

  get meals() {
    return [...this.data.meals];
  }

  get warnings() {
    return [...this.data.warnings];
  }

  get templateWeek() {
    return [...this.data.templateWeek];
  }

  set id(scheduleId) {
    this.data.id = scheduleId || null;
  }

  set defaultServings(defaultServings) {
    this.data.defaultServings = defaultServings || 2;
    this.updateTemplateWeek();
  }

  set strategy(strategy) {
    this.data.strategy = strategy;
    this.updateTemplateWeek();
  }

  getScheduleMeal = (day: number, meal: number): ScheduleMeal => {
    return this.meals[day * 3 + meal];
  };

  setScheduleMealDetails = (dayIndex, mealIndex, details = {}) => {
    const meal = this.data.meals[dayIndex * 3 + mealIndex];
    Object.assign(meal, details);
    this.updateTemplateWeek();
  };

  updateTemplateWeek = () => {
    this.data.templateWeek = planner.run(this);
  };

  getPlanMeal = (dateIndex: number, mealIndex: number) => {
    // copy basic data from template week and update date
    const dayIndex = dateIndex % 7;
    const templateMeal = this.data.templateWeek.find(
      meal => meal.dayIndex === dayIndex && meal.mealIndex === mealIndex,
    );
    return templateMeal.move(Math.floor((dateIndex - dayIndex) / 7));
  };

  addWarning = warning => {
    this.data.warnings.push(warning);
  };

  reset = () => {
    this.data.warnings = [];
  };

  countMealsWithAvailability = availability => {
    const options = [].concat(availability);
    return this.meals.filter(meal => options.includes(meal.availability))
      .length;
  };

  toJSON() {
    return {
      ...this.data,
      templateWeek: this.data.templateWeek.map(meal => meal.toJSON()),
    };
  }

  static fromJSON(data) {
    return new Schedule(data);
  }

  static createEmpty(id: string) {
    return new Schedule({
      id,
      defaultServings: 2,
      strategy: ScheduleStrategy.Basic,

      meals: new Array(21).fill(null).map((__, mealIndex) => ({
        id: `${id}_m_${mealIndex}`,
        mealIndex,
        dayIndex: Math.floor(mealIndex / 3),
        availability: 'SKIP',
        servings: 2,
      })),

      templateWeek: new Array(21)
        .fill(null)
        .map((_, mealIndex) =>
          Meal.createEmpty(Math.floor(mealIndex / 3), mealIndex % 3),
        ),

      warnings: [],
    });
  }
}
