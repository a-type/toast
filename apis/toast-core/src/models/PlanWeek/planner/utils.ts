import Schedule, { ScheduleMeal } from 'models/Schedule';
import { PlanWeekData } from '../types';

const MEALS_PER_DAY = 3;
const DAYS_PER_WEEK = 7;

export const mealsFromGroceryDay = (schedule: Schedule): ScheduleMeal[] => {
  return [
    ...schedule.meals.slice(schedule.groceryDay * MEALS_PER_DAY),
    ...schedule.meals.slice(0, schedule.groceryDay * MEALS_PER_DAY),
  ];
};

const findCookedMeal = meal =>
  meal.mealIndex !== 0 &&
  ['LONG', 'MEDIUM', 'SHORT'].includes(meal.availability);

export const initializeWeek = (
  schedule: Schedule,
  weekIndex: number,
  findFirstMeal: (meal: ScheduleMeal) => boolean = findCookedMeal,
) => {
  // skip breakfast for main planning
  // find first cook day after grocery day to use to initialize the week
  const firstCookMeal = mealsFromGroceryDay(schedule).find(findFirstMeal);

  if (!firstCookMeal) {
    throw new Error(
      "Sorry, we couldn't figure out where to start this plan based on your schedule.",
    );
  }

  const week: PlanWeekData = {
    id: `${schedule.id}_w_${weekIndex}`,
    weekIndex,
    mealsPerDay: MEALS_PER_DAY,
    startDay: firstCookMeal.dayIndex,
    startMeal: firstCookMeal.mealIndex,
    meals: new Array(MEALS_PER_DAY * DAYS_PER_WEEK)
      .fill(null)
      .map((_, idx) => ({
        id: `${schedule.id}_w_${weekIndex}_m_${idx}`,
        weekIndex,
        // offsets meal index according to the meal the week starts on
        mealIndex: (firstCookMeal.mealIndex + idx) % MEALS_PER_DAY,
        dayIndex: Math.floor(
          ((firstCookMeal.dayIndex * MEALS_PER_DAY +
            firstCookMeal.mealIndex +
            idx) /
            MEALS_PER_DAY) %
            DAYS_PER_WEEK,
        ),
        actions: [],
      })),
  };

  return week;
};
