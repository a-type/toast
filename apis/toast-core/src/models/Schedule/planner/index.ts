import basic from './basic';
import prep from './prep';
import { Schedule, Meal } from 'models';
import { ScheduleStrategy } from 'models/Schedule/Schedule';
import { initializeWeek } from './utils';

export default {
  run: (schedule: Schedule): Meal[] => {
    const meals = initializeWeek();

    switch (schedule.strategy) {
      case ScheduleStrategy.Prep:
        return prep(schedule, meals);
      default:
        return basic(schedule, meals);
    }
  },
};
