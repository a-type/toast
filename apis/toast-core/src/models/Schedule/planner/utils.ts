import { ScheduleMeal } from 'models/Schedule/Schedule';
import { Meal, Schedule } from 'models';

export const MEALS_PER_DAY = 3;
export const DAYS_PER_WEEK = 7;

export const mealsFromGroceryDay = (schedule: Schedule): ScheduleMeal[] => {
  return [
    ...schedule.meals.slice(schedule.groceryDay * MEALS_PER_DAY),
    ...schedule.meals.slice(0, schedule.groceryDay * MEALS_PER_DAY),
  ];
};

const findCookedMeal = meal =>
  meal.mealIndex !== 0 &&
  ['LONG', 'MEDIUM', 'SHORT'].includes(meal.availability);

export const initializeWeek = (
  schedule: Schedule,
  findFirstMeal: (meal: ScheduleMeal) => boolean = findCookedMeal,
): Meal[] => {
  // skip breakfast for main planning
  // find first cook day after grocery day to use to initialize the week
  const fromGroceryDay = mealsFromGroceryDay(schedule);
  const firstCookMeal = fromGroceryDay.find(findFirstMeal) || fromGroceryDay[0];

  const meals = new Array(DAYS_PER_WEEK)
    .fill(null)
    .map((_, dayIndex) =>
      new Array(MEALS_PER_DAY).fill(null).map((__, baseMealIndex) => {
        const dateIndex = (dayIndex + firstCookMeal.dayIndex) % 7;
        const mealIndex = (baseMealIndex + firstCookMeal.mealIndex) % 3;

        return Meal.createEmpty(dateIndex, mealIndex);
      }),
    )
    .reduce((allMeals, dayMeals) => allMeals.concat(dayMeals), []);

  return meals;
};
