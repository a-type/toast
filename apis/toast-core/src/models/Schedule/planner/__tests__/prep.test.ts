import plan from '../index';
import { ScheduleStrategy } from 'models/Schedule/Schedule';
import plans from './createPlans';
import { ignoreIds } from './utils';
const {
  variety,
  largeFamily,
  busy,
  nefarious,
  basic,
  midweekGrocery,
} = plans();

describe('prep meal plan', () => {
  test('for a variety plan', () => {
    variety.strategy = ScheduleStrategy.Prep;
    expect(ignoreIds(plan.run(variety).map(m => m.toJSON()))).toMatchSnapshot();
  });

  test('for midweek grocery day', () => {
    midweekGrocery.strategy = ScheduleStrategy.Prep;
    expect(
      ignoreIds(plan.run(midweekGrocery).map(m => m.toJSON())),
    ).toMatchSnapshot();
  });

  test('for a large family plan', () => {
    largeFamily.strategy = ScheduleStrategy.Prep;
    expect(
      ignoreIds(plan.run(largeFamily).map(m => m.toJSON())),
    ).toMatchSnapshot();
  });

  test('for a really busy person', () => {
    busy.strategy = ScheduleStrategy.Prep;
    expect(ignoreIds(plan.run(busy).map(m => m.toJSON()))).toMatchSnapshot();
  });

  test('a basic plan', () => {
    basic.strategy = ScheduleStrategy.Prep;
    expect(ignoreIds(plan.run(basic).map(m => m.toJSON()))).toMatchSnapshot();
  });

  describe('nefarious cases', () => {
    test('no time', () => {
      nefarious.noTime.strategy = ScheduleStrategy.Prep;
      expect(() =>
        ignoreIds(plan.run(nefarious.noTime).map(m => m.toJSON())),
      ).toMatchSnapshot();
    });

    test('eating out all the time', () => {
      nefarious.eatOut.strategy = ScheduleStrategy.Prep;
      expect(() =>
        ignoreIds(plan.run(nefarious.eatOut).map(m => m.toJSON())),
      ).toMatchSnapshot();
    });
  });
});
