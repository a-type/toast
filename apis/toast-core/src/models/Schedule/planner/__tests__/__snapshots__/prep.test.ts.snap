// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`prep meal plan a basic plan 1`] = `
Array [
  Object {
    "actions": Array [
      Object {
        "mealType": "NORMAL",
        "servings": 4,
        "type": "COOK",
      },
      Object {
        "leftovers": false,
        "type": "EAT",
      },
    ],
    "dateIndex": 0,
    "mealIndex": 1,
  },
  Object {
    "actions": Array [
      Object {
        "mealType": "FANCY",
        "servings": 2,
        "type": "COOK",
      },
      Object {
        "leftovers": false,
        "type": "EAT",
      },
    ],
    "dateIndex": 0,
    "mealIndex": 2,
  },
  Object {
    "actions": Array [],
    "dateIndex": 0,
    "mealIndex": 0,
  },
  Object {
    "actions": Array [
      Object {
        "dayIndex": 1,
        "leftovers": true,
        "mealIndex": 1,
        "type": "EAT",
      },
    ],
    "dateIndex": 1,
    "mealIndex": 1,
  },
  Object {
    "actions": Array [
      Object {
        "mealType": "FANCY",
        "servings": 4,
        "type": "COOK",
      },
      Object {
        "leftovers": false,
        "type": "EAT",
      },
    ],
    "dateIndex": 1,
    "mealIndex": 2,
  },
  Object {
    "actions": Array [],
    "dateIndex": 1,
    "mealIndex": 0,
  },
  Object {
    "actions": Array [
      Object {
        "dayIndex": 2,
        "leftovers": true,
        "mealIndex": 1,
        "type": "EAT",
      },
    ],
    "dateIndex": 2,
    "mealIndex": 1,
  },
  Object {
    "actions": Array [
      Object {
        "mealType": "FANCY",
        "servings": 4,
        "type": "COOK",
      },
      Object {
        "leftovers": false,
        "type": "EAT",
      },
    ],
    "dateIndex": 2,
    "mealIndex": 2,
  },
  Object {
    "actions": Array [],
    "dateIndex": 2,
    "mealIndex": 0,
  },
  Object {
    "actions": Array [
      Object {
        "dayIndex": 3,
        "leftovers": true,
        "mealIndex": 1,
        "type": "EAT",
      },
    ],
    "dateIndex": 3,
    "mealIndex": 1,
  },
  Object {
    "actions": Array [
      Object {
        "mealType": "NORMAL",
        "servings": 6,
        "type": "COOK",
      },
      Object {
        "leftovers": false,
        "type": "EAT",
      },
    ],
    "dateIndex": 3,
    "mealIndex": 2,
  },
  Object {
    "actions": Array [],
    "dateIndex": 3,
    "mealIndex": 0,
  },
  Object {
    "actions": Array [
      Object {
        "dayIndex": 4,
        "leftovers": true,
        "mealIndex": 1,
        "type": "EAT",
      },
    ],
    "dateIndex": 4,
    "mealIndex": 1,
  },
  Object {
    "actions": Array [
      Object {
        "mealType": "QUICK",
        "servings": 2,
        "type": "COOK",
      },
      Object {
        "leftovers": false,
        "type": "EAT",
      },
    ],
    "dateIndex": 4,
    "mealIndex": 2,
  },
  Object {
    "actions": Array [],
    "dateIndex": 4,
    "mealIndex": 0,
  },
  Object {
    "actions": Array [
      Object {
        "dayIndex": 5,
        "leftovers": true,
        "mealIndex": 1,
        "type": "EAT",
      },
    ],
    "dateIndex": 5,
    "mealIndex": 1,
  },
  Object {
    "actions": Array [
      Object {
        "mealType": "QUICK",
        "servings": 2,
        "type": "COOK",
      },
      Object {
        "leftovers": false,
        "type": "EAT",
      },
    ],
    "dateIndex": 5,
    "mealIndex": 2,
  },
  Object {
    "actions": Array [],
    "dateIndex": 5,
    "mealIndex": 0,
  },
  Object {
    "actions": Array [
      Object {
        "mealType": "NORMAL",
        "servings": 2,
        "type": "COOK",
      },
      Object {
        "leftovers": false,
        "type": "EAT",
      },
    ],
    "dateIndex": 6,
    "mealIndex": 1,
  },
  Object {
    "actions": Array [
      Object {
        "mealType": "FANCY",
        "servings": 2,
        "type": "COOK",
      },
      Object {
        "leftovers": false,
        "type": "EAT",
      },
    ],
    "dateIndex": 6,
    "mealIndex": 2,
  },
  Object {
    "actions": Array [],
    "dateIndex": 6,
    "mealIndex": 0,
  },
]
`;

exports[`prep meal plan for a large family plan 1`] = `
Array [
  Object {
    "actions": Array [
      Object {
        "mealType": "BIG",
        "servings": 12,
        "type": "COOK",
      },
      Object {
        "leftovers": false,
        "type": "EAT",
      },
    ],
    "dateIndex": 0,
    "mealIndex": 2,
  },
  Object {
    "actions": Array [],
    "dateIndex": 0,
    "mealIndex": 0,
  },
  Object {
    "actions": Array [
      Object {
        "type": "EAT_OUT",
      },
    ],
    "dateIndex": 0,
    "mealIndex": 1,
  },
  Object {
    "actions": Array [
      Object {
        "mealType": "FANCY",
        "servings": 6,
        "type": "COOK",
      },
      Object {
        "leftovers": false,
        "type": "EAT",
      },
    ],
    "dateIndex": 1,
    "mealIndex": 2,
  },
  Object {
    "actions": Array [],
    "dateIndex": 1,
    "mealIndex": 0,
  },
  Object {
    "actions": Array [
      Object {
        "mealType": "QUICK",
        "servings": 6,
        "type": "COOK",
      },
      Object {
        "leftovers": false,
        "type": "EAT",
      },
    ],
    "dateIndex": 1,
    "mealIndex": 1,
  },
  Object {
    "actions": Array [
      Object {
        "mealType": "QUICK",
        "servings": 6,
        "type": "COOK",
      },
      Object {
        "leftovers": false,
        "type": "EAT",
      },
    ],
    "dateIndex": 2,
    "mealIndex": 2,
  },
  Object {
    "actions": Array [],
    "dateIndex": 2,
    "mealIndex": 0,
  },
  Object {
    "actions": Array [
      Object {
        "mealType": "QUICK",
        "servings": 6,
        "type": "COOK",
      },
      Object {
        "leftovers": false,
        "type": "EAT",
      },
    ],
    "dateIndex": 2,
    "mealIndex": 1,
  },
  Object {
    "actions": Array [
      Object {
        "type": "EAT_OUT",
      },
    ],
    "dateIndex": 3,
    "mealIndex": 2,
  },
  Object {
    "actions": Array [],
    "dateIndex": 3,
    "mealIndex": 0,
  },
  Object {
    "actions": Array [
      Object {
        "dayIndex": 3,
        "leftovers": true,
        "mealIndex": 1,
        "type": "EAT",
      },
    ],
    "dateIndex": 3,
    "mealIndex": 1,
  },
  Object {
    "actions": Array [
      Object {
        "mealType": "BIG",
        "servings": 12,
        "type": "COOK",
      },
      Object {
        "leftovers": false,
        "type": "EAT",
      },
    ],
    "dateIndex": 4,
    "mealIndex": 2,
  },
  Object {
    "actions": Array [],
    "dateIndex": 4,
    "mealIndex": 0,
  },
  Object {
    "actions": Array [
      Object {
        "mealType": "QUICK",
        "servings": 6,
        "type": "COOK",
      },
      Object {
        "leftovers": false,
        "type": "EAT",
      },
    ],
    "dateIndex": 4,
    "mealIndex": 1,
  },
  Object {
    "actions": Array [
      Object {
        "mealType": "NORMAL",
        "servings": 6,
        "type": "COOK",
      },
      Object {
        "leftovers": false,
        "type": "EAT",
      },
    ],
    "dateIndex": 5,
    "mealIndex": 2,
  },
  Object {
    "actions": Array [],
    "dateIndex": 5,
    "mealIndex": 0,
  },
  Object {
    "actions": Array [
      Object {
        "dayIndex": 5,
        "leftovers": true,
        "mealIndex": 1,
        "type": "EAT",
      },
    ],
    "dateIndex": 5,
    "mealIndex": 1,
  },
  Object {
    "actions": Array [
      Object {
        "mealType": "BIG",
        "servings": 12,
        "type": "COOK",
      },
      Object {
        "leftovers": false,
        "type": "EAT",
      },
    ],
    "dateIndex": 6,
    "mealIndex": 2,
  },
  Object {
    "actions": Array [],
    "dateIndex": 6,
    "mealIndex": 0,
  },
  Object {
    "actions": Array [
      Object {
        "dayIndex": 6,
        "leftovers": true,
        "mealIndex": 1,
        "type": "EAT",
      },
    ],
    "dateIndex": 6,
    "mealIndex": 1,
  },
]
`;

exports[`prep meal plan for a really busy person 1`] = `
Array [
  Object {
    "actions": Array [
      Object {
        "mealType": "QUICK",
        "servings": 2,
        "type": "COOK",
      },
      Object {
        "leftovers": false,
        "type": "EAT",
      },
    ],
    "dateIndex": 0,
    "mealIndex": 1,
  },
  Object {
    "actions": Array [
      Object {
        "mealType": "QUICK",
        "servings": 2,
        "type": "COOK",
      },
      Object {
        "leftovers": false,
        "type": "EAT",
      },
    ],
    "dateIndex": 0,
    "mealIndex": 2,
  },
  Object {
    "actions": Array [],
    "dateIndex": 0,
    "mealIndex": 0,
  },
  Object {
    "actions": Array [
      Object {
        "type": "EAT_OUT",
      },
    ],
    "dateIndex": 1,
    "mealIndex": 1,
  },
  Object {
    "actions": Array [],
    "dateIndex": 1,
    "mealIndex": 2,
  },
  Object {
    "actions": Array [],
    "dateIndex": 1,
    "mealIndex": 0,
  },
  Object {
    "actions": Array [],
    "dateIndex": 2,
    "mealIndex": 1,
  },
  Object {
    "actions": Array [
      Object {
        "mealType": "QUICK",
        "servings": 2,
        "type": "COOK",
      },
      Object {
        "leftovers": false,
        "type": "EAT",
      },
    ],
    "dateIndex": 2,
    "mealIndex": 2,
  },
  Object {
    "actions": Array [],
    "dateIndex": 2,
    "mealIndex": 0,
  },
  Object {
    "actions": Array [],
    "dateIndex": 3,
    "mealIndex": 1,
  },
  Object {
    "actions": Array [
      Object {
        "type": "EAT_OUT",
      },
    ],
    "dateIndex": 3,
    "mealIndex": 2,
  },
  Object {
    "actions": Array [],
    "dateIndex": 3,
    "mealIndex": 0,
  },
  Object {
    "actions": Array [],
    "dateIndex": 4,
    "mealIndex": 1,
  },
  Object {
    "actions": Array [
      Object {
        "mealType": "QUICK",
        "servings": 8,
        "type": "COOK",
      },
      Object {
        "leftovers": false,
        "type": "EAT",
      },
    ],
    "dateIndex": 4,
    "mealIndex": 2,
  },
  Object {
    "actions": Array [],
    "dateIndex": 4,
    "mealIndex": 0,
  },
  Object {
    "actions": Array [
      Object {
        "dayIndex": 5,
        "leftovers": true,
        "mealIndex": 1,
        "type": "EAT",
      },
    ],
    "dateIndex": 5,
    "mealIndex": 1,
  },
  Object {
    "actions": Array [
      Object {
        "mealType": "QUICK",
        "servings": 2,
        "type": "COOK",
      },
      Object {
        "leftovers": false,
        "type": "EAT",
      },
    ],
    "dateIndex": 5,
    "mealIndex": 2,
  },
  Object {
    "actions": Array [],
    "dateIndex": 5,
    "mealIndex": 0,
  },
  Object {
    "actions": Array [
      Object {
        "dayIndex": 6,
        "leftovers": true,
        "mealIndex": 1,
        "type": "EAT",
      },
    ],
    "dateIndex": 6,
    "mealIndex": 1,
  },
  Object {
    "actions": Array [
      Object {
        "dayIndex": 6,
        "leftovers": true,
        "mealIndex": 2,
        "type": "EAT",
      },
    ],
    "dateIndex": 6,
    "mealIndex": 2,
  },
  Object {
    "actions": Array [],
    "dateIndex": 6,
    "mealIndex": 0,
  },
]
`;

exports[`prep meal plan for a variety plan 1`] = `
Array [
  Object {
    "actions": Array [
      Object {
        "mealType": "QUICK",
        "servings": 2,
        "type": "COOK",
      },
      Object {
        "leftovers": false,
        "type": "EAT",
      },
    ],
    "dateIndex": 0,
    "mealIndex": 1,
  },
  Object {
    "actions": Array [
      Object {
        "mealType": "FANCY",
        "servings": 4,
        "type": "COOK",
      },
      Object {
        "leftovers": false,
        "type": "EAT",
      },
    ],
    "dateIndex": 0,
    "mealIndex": 2,
  },
  Object {
    "actions": Array [],
    "dateIndex": 0,
    "mealIndex": 0,
  },
  Object {
    "actions": Array [
      Object {
        "type": "EAT_OUT",
      },
    ],
    "dateIndex": 1,
    "mealIndex": 1,
  },
  Object {
    "actions": Array [
      Object {
        "mealType": "NORMAL",
        "servings": 4,
        "type": "COOK",
      },
      Object {
        "leftovers": false,
        "type": "EAT",
      },
    ],
    "dateIndex": 1,
    "mealIndex": 2,
  },
  Object {
    "actions": Array [],
    "dateIndex": 1,
    "mealIndex": 0,
  },
  Object {
    "actions": Array [
      Object {
        "mealType": "QUICK",
        "servings": 2,
        "type": "COOK",
      },
      Object {
        "leftovers": false,
        "type": "EAT",
      },
    ],
    "dateIndex": 2,
    "mealIndex": 1,
  },
  Object {
    "actions": Array [
      Object {
        "mealType": "QUICK",
        "servings": 2,
        "type": "COOK",
      },
      Object {
        "leftovers": false,
        "type": "EAT",
      },
    ],
    "dateIndex": 2,
    "mealIndex": 2,
  },
  Object {
    "actions": Array [],
    "dateIndex": 2,
    "mealIndex": 0,
  },
  Object {
    "actions": Array [
      Object {
        "dayIndex": 3,
        "leftovers": true,
        "mealIndex": 1,
        "type": "EAT",
      },
    ],
    "dateIndex": 3,
    "mealIndex": 1,
  },
  Object {
    "actions": Array [
      Object {
        "type": "EAT_OUT",
      },
    ],
    "dateIndex": 3,
    "mealIndex": 2,
  },
  Object {
    "actions": Array [],
    "dateIndex": 3,
    "mealIndex": 0,
  },
  Object {
    "actions": Array [
      Object {
        "dayIndex": 4,
        "leftovers": true,
        "mealIndex": 1,
        "type": "EAT",
      },
    ],
    "dateIndex": 4,
    "mealIndex": 1,
  },
  Object {
    "actions": Array [
      Object {
        "mealType": "NORMAL",
        "servings": 4,
        "type": "COOK",
      },
      Object {
        "leftovers": false,
        "type": "EAT",
      },
    ],
    "dateIndex": 4,
    "mealIndex": 2,
  },
  Object {
    "actions": Array [],
    "dateIndex": 4,
    "mealIndex": 0,
  },
  Object {
    "actions": Array [
      Object {
        "mealType": "QUICK",
        "servings": 2,
        "type": "COOK",
      },
      Object {
        "leftovers": false,
        "type": "EAT",
      },
    ],
    "dateIndex": 5,
    "mealIndex": 1,
  },
  Object {
    "actions": Array [
      Object {
        "mealType": "FANCY",
        "servings": 2,
        "type": "COOK",
      },
      Object {
        "leftovers": false,
        "type": "EAT",
      },
    ],
    "dateIndex": 5,
    "mealIndex": 2,
  },
  Object {
    "actions": Array [],
    "dateIndex": 5,
    "mealIndex": 0,
  },
  Object {
    "actions": Array [
      Object {
        "dayIndex": 6,
        "leftovers": true,
        "mealIndex": 1,
        "type": "EAT",
      },
    ],
    "dateIndex": 6,
    "mealIndex": 1,
  },
  Object {
    "actions": Array [
      Object {
        "mealType": "FANCY",
        "servings": 2,
        "type": "COOK",
      },
      Object {
        "leftovers": false,
        "type": "EAT",
      },
    ],
    "dateIndex": 6,
    "mealIndex": 2,
  },
  Object {
    "actions": Array [],
    "dateIndex": 6,
    "mealIndex": 0,
  },
]
`;

exports[`prep meal plan for midweek grocery day 1`] = `
Array [
  Object {
    "actions": Array [
      Object {
        "mealType": "NORMAL",
        "servings": 4,
        "type": "COOK",
      },
      Object {
        "leftovers": false,
        "type": "EAT",
      },
    ],
    "dateIndex": 4,
    "mealIndex": 2,
  },
  Object {
    "actions": Array [],
    "dateIndex": 4,
    "mealIndex": 0,
  },
  Object {
    "actions": Array [
      Object {
        "dayIndex": 4,
        "leftovers": true,
        "mealIndex": 1,
        "type": "EAT",
      },
    ],
    "dateIndex": 4,
    "mealIndex": 1,
  },
  Object {
    "actions": Array [
      Object {
        "mealType": "FANCY",
        "servings": 4,
        "type": "COOK",
      },
      Object {
        "leftovers": false,
        "type": "EAT",
      },
    ],
    "dateIndex": 5,
    "mealIndex": 2,
  },
  Object {
    "actions": Array [],
    "dateIndex": 5,
    "mealIndex": 0,
  },
  Object {
    "actions": Array [
      Object {
        "mealType": "QUICK",
        "servings": 2,
        "type": "COOK",
      },
      Object {
        "leftovers": false,
        "type": "EAT",
      },
    ],
    "dateIndex": 5,
    "mealIndex": 1,
  },
  Object {
    "actions": Array [
      Object {
        "mealType": "FANCY",
        "servings": 2,
        "type": "COOK",
      },
      Object {
        "leftovers": false,
        "type": "EAT",
      },
    ],
    "dateIndex": 6,
    "mealIndex": 2,
  },
  Object {
    "actions": Array [],
    "dateIndex": 6,
    "mealIndex": 0,
  },
  Object {
    "actions": Array [
      Object {
        "dayIndex": 6,
        "leftovers": true,
        "mealIndex": 1,
        "type": "EAT",
      },
    ],
    "dateIndex": 6,
    "mealIndex": 1,
  },
  Object {
    "actions": Array [
      Object {
        "mealType": "FANCY",
        "servings": 4,
        "type": "COOK",
      },
      Object {
        "leftovers": false,
        "type": "EAT",
      },
    ],
    "dateIndex": 0,
    "mealIndex": 2,
  },
  Object {
    "actions": Array [],
    "dateIndex": 0,
    "mealIndex": 0,
  },
  Object {
    "actions": Array [
      Object {
        "mealType": "QUICK",
        "servings": 2,
        "type": "COOK",
      },
      Object {
        "leftovers": false,
        "type": "EAT",
      },
    ],
    "dateIndex": 0,
    "mealIndex": 1,
  },
  Object {
    "actions": Array [
      Object {
        "mealType": "NORMAL",
        "servings": 2,
        "type": "COOK",
      },
      Object {
        "leftovers": false,
        "type": "EAT",
      },
    ],
    "dateIndex": 1,
    "mealIndex": 2,
  },
  Object {
    "actions": Array [],
    "dateIndex": 1,
    "mealIndex": 0,
  },
  Object {
    "actions": Array [
      Object {
        "type": "EAT_OUT",
      },
    ],
    "dateIndex": 1,
    "mealIndex": 1,
  },
  Object {
    "actions": Array [
      Object {
        "mealType": "QUICK",
        "servings": 2,
        "type": "COOK",
      },
      Object {
        "leftovers": false,
        "type": "EAT",
      },
    ],
    "dateIndex": 2,
    "mealIndex": 2,
  },
  Object {
    "actions": Array [],
    "dateIndex": 2,
    "mealIndex": 0,
  },
  Object {
    "actions": Array [
      Object {
        "mealType": "QUICK",
        "servings": 2,
        "type": "COOK",
      },
      Object {
        "leftovers": false,
        "type": "EAT",
      },
    ],
    "dateIndex": 2,
    "mealIndex": 1,
  },
  Object {
    "actions": Array [
      Object {
        "type": "EAT_OUT",
      },
    ],
    "dateIndex": 3,
    "mealIndex": 2,
  },
  Object {
    "actions": Array [],
    "dateIndex": 3,
    "mealIndex": 0,
  },
  Object {
    "actions": Array [
      Object {
        "dayIndex": 3,
        "leftovers": true,
        "mealIndex": 1,
        "type": "EAT",
      },
    ],
    "dateIndex": 3,
    "mealIndex": 1,
  },
]
`;

exports[`prep meal plan nefarious cases eating out all the time 1`] = `[Function]`;

exports[`prep meal plan nefarious cases no time 1`] = `[Function]`;
